### PR TITLE
Internal pid filtering for STREAM input

### PR DESCRIPTION
**Pull request by lars18th**
Implementation of the internal pid filtering for the STREAM input. Now the writing of the packets isn't packed. Therefore if some (but not all) packets are filtered, then the writed buffer is less than the common 7 TS packets. This strategy isn't optimal but it works without errors.

### CODE REVIEW

**Comment by lars18th** [comment]
Hi @Barracuda09 ,

This is my first implementation for the STREAM input. Is not optimal because it flush all incomplete packets. However, it's fully functional. Futhermore, it illustrates the changes required to port this functionality to other devices. I'll try next for (FILE) TS Reader. And I'll optimize them after. You agree to merge it?

File: src/input/stream/Streamer.cpp
Comment (Barracuda09) [inline src/input/stream/Streamer.cpp:111]: Is this do-while loop constuction needed?

I still fail to understand the `buffer.markToFlush()`, I know you said it is more efficient.

I think this could be written as (you agree?):
```
  if (readSize > 0) {
    buffer.addAmountOfBytesWritten(readSize);
    buffer.trySyncing();
    // Add data to Filter
    _deviceData.getFilterData().addData(_feID, buffer, _deviceData.isInternalPidFilteringEnabled());
    return buffer.full();
  }
  return false
```
This is more straight forward and the buffer will be filled on next run as it always was.

**Review by Barracuda09 — CHANGES_REQUESTED** [review]
Hi @lars18th 

Please look at my feedback,

File: src/input/stream/Streamer.cpp
Comment (lars18th) [reply src/input/stream/Streamer.cpp:130]: Hi @Barracuda09 ,

Sorry! You're right. This is a typo (error in my merge). Fixed now.

File: src/input/stream/StreamerData.cpp
Comment (lars18th) [reply src/input/stream/StreamerData.cpp:60]: Hi @Barracuda09 ,

My idea is to _improve_ this after you merge this PR. If you want, after I'll change this name to another different. I don't want to do now beacuase is the same name used inside the Child PIPE for the same functionality. And my objective is to promote these functions to the Base class as virtual with a default behaviour. However, even if you merge it I'll save this comment for my next PR. You agree?

File: src/input/stream/Streamer.cpp
Comment (lars18th) [reply src/input/stream/Streamer.cpp:111]: Hi @Barracuda09 ,

Yes, I feel the problem is that you don't get the reason for the function `buffer.markToFlush()`.
So please check this code in the current master:
https://github.com/Barracuda09/SATPI/blob/master/src/output/StreamThreadBase.cpp#L226-L234
This is executed in `StreamThreadBase::readDataFromInputDevice()` after calling to `inputDevice->readFullTSPacket(_tsBuffer[_writeIndex])`. And if the `_tsBuffer` not pass the check of `isReadyToSend()` then it never will be sended but discarded. Or I'm missing something?

Different from that, my block of `do { ... } while (1)` is becuase I want to fully read the Buffer (7 TS packets). If not then, I can re-read or flush it. The current implementation is based on the same code as for Child PIPE. However, I'll change it in the future when I'll add a re-read if more data is waiting in the incoming socket. I hope you agree with this.

But the main question is what happens with a non-full Buffer? If the buffer will wait for future reading, then I can reduce the complexity. But the check of `isReadyToSend()` will remove the data (empty the buffer) before returning to calling to `StreamThreadBase::readDataFromInputDevice()`, what I don't want. Any idea to fix or improve this?

Comment (Barracuda09) [reply src/input/stream/Streamer.cpp:111]: > https://github.com/Barracuda09/SATPI/blob/master/src/output/StreamThreadBase.cpp#L226-L234

This is only to send a empty (null) packet if the buffer is not full in 100 microseconds. So this should be no problem.

**Review by Barracuda09 — COMMENTED** [review]
https://github.com/Barracuda09/SATPI/pull/168/files#r1013310441

**Comment by lars18th** [comment]
Hi @Barracuda09 ,

Thank you for merging it. And I agree with your changes. However, I'll provide more refactorings. Please remain on hold. 😉

**Pull request merged**
